src/Vault.sol

```diff
@@ -768,16 +935,104 @@ contract Vault {
         }
     }
 
-    function onERC721Received(address, address, uint256, bytes calldata) public pure returns (bytes4) {
+    function onERC721Received(
+        address,
+        address,
+        uint256,
+        bytes calldata
+    ) public pure returns (bytes4) {
         return this.onERC721Received.selector;
     }
 
-    function onERC1155Received(address, address, uint256, uint256, bytes calldata) public pure returns (bytes4) {
+    function onERC1155Received(
+        address,
+        address,
+        uint256,
+        uint256,
+        bytes calldata
+    ) public pure returns (bytes4) {
         return this.onERC1155Received.selector;
     }
 
     //Function only used for tests
-    function getLengths() external view returns (uint256, uint256, uint256, uint256) {
-        return (erc20Stored.length, erc721Stored.length, erc721TokenIds.length, erc1155Stored.length);
+    function getLengths()
+        external
+        view
+        returns (
+            uint256,
+            uint256,
+            uint256,
+            uint256
+        )
+    {
+        return (
+            erc20Stored.length,
+            erc721Stored.length,
+            erc721TokenIds.length,
+            erc1155Stored.length
+        );
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                    ASSET MANAGEMENT LOGIC
+    ///////////////////////////////////////////////////////////////*/
+
+    function vaultManagementAction(
+        address _actionHandler,
+        bytes calldata _actionData
+    ) public onlyOwner {
+        // causes weird reverts
+        require(
+            IMainRegistry(registryAddress).isActionAllowlisted(_actionHandler),
+            "VL_VMA: Action is not allowlisted"
+        );
+
+        (
+            actionAssetsData memory outgoing_,
+            actionAssetsData memory incoming_,
+
+        ) = abi.decode(
+                _actionData,
+                (actionAssetsData, actionAssetsData, bytes)
+            );
+
+        //TODO: account preAction CT
+
+        // account preActionBalances
+
+        for (uint256 i; i < incoming_.assets.length; i++) {
+            incoming_.preActionBalances[i] = IERC20(incoming_.assets[i])
+                .balanceOf(address(this));
+        }
+
+        // withdraw to actionHandler
+        for (uint256 i; i < outgoing_.assets.length; i++) {
+            outgoing_.preActionBalances[i] = IERC20(outgoing_.assets[i])
+                .balanceOf(address(this));
+            _withdrawERC20(
+                _actionHandler,
+                outgoing_.assets[i],
+                outgoing_.assetAmounts[i]
+            );
+        }
+
+        // execute Action
+        actionAssetsData memory _incomingAssets = IActionBase(_actionHandler)
+            .executeAction(address(this), msg.sender, _actionData);
+
+        // deposit back into vault (approval in Action)
+        for (uint256 i; i < incoming_.assets.length; i++) {
+            _depositERC20(
```

Comment:
This will not work, tokens are not sent to the actionhandler by the swap! check why this isn't captured in the tests!